src/QueryReflection/PdoQueryReflector.php

```diff
@@ -0,0 +1,231 @@
+    private function simulateQuery(string $queryString)
+    {
+        if (\array_key_exists($queryString, $this->cache)) {
+            return $this->cache[$queryString];
+        }
+
+        if (\count($this->cache) > self::MAX_CACHE_SIZE) {
+            // make room for the next element by randomly removing a existing one
+            array_shift($this->cache);
+        }
+
+        $simulatedQuery = QuerySimulation::simulate($queryString);
+        if (null === $simulatedQuery) {
+            return $this->cache[$queryString] = null;
+        }
+
+        try {
+            $result = $this->pdo->query($simulatedQuery);
+        } catch (PDOException $e) {
+            return $this->cache[$queryString] = $e;
+        }
+
+        $this->cache[$queryString] = [];
+        $columnCount = $result->columnCount();
+        $columnIndex = 0;
+        while ($columnIndex < $columnCount) {
+            // see https://github.com/php/php-src/blob/master/ext/pdo_mysql/mysql_statement.c
+            $columnMeta = $result->getColumnMeta($columnIndex);
+
+            if (false === $columnMeta) {
+                throw new ShouldNotHappenException('Failed to get column meta for column index '.$columnIndex);
+            }
+
+            if (
+                !array_key_exists('name', $columnMeta)
+                || !array_key_exists('table', $columnMeta)
+                || !array_key_exists('native_type', $columnMeta)
+                || !array_key_exists('flags', $columnMeta)
+                || !array_key_exists('len', $columnMeta)
+            ) {
+                throw new ShouldNotHappenException();
+            }
+
+            if ($this->isAutoIncrementCol($columnMeta['table'], $columnMeta['name'])) {
+                $columnMeta['flags'][] = MysqlTypeMapper::FLAG_AUTO_INCREMENT;
+            }
+            if ($this->isNumericCol($columnMeta['native_type'])) {
+                $columnMeta['flags'][] = MysqlTypeMapper::FLAG_NUMERIC;
+            }
+
+            // @phpstan-ignore-next-line
+            $this->cache[$queryString][$columnIndex] = $columnMeta;
+            ++$columnIndex;
+        }
+
+        // @phpstan-ignore-next-line
+        return $this->cache[$queryString];
+    }
+
+    private function isAutoIncrementCol(string $tableName, string $columnName):bool {
+        if (array_key_exists($tableName, $this->autoIncrementColumns)) {
+            return array_key_exists($columnName, $this->autoIncrementColumns[$tableName]);
+        }
+
+        if ($this->stmt === null) {
+            $this->stmt = $this->pdo->prepare(
+                'SELECT EXTRA FROM information_schema.columns WHERE table_name = ? AND table_schema = DATABASE()'
+            );
+        }
+        $this->stmt->execute([$tableName]);
+        $result = $this->stmt->fetchAll(PDO::FETCH_ASSOC);
+
+        $this->autoIncrementColumns[$tableName] = [];
+        foreach($result as $row) {
+            $extra = $row['EXTRA'] ?? '';
```

Comment:
We might fix the

> Offset 'EXTRA' on array{EXTRA: string} on left side of ?? always exists and is not nullable.

error by using NVL within the query. I guess mariadb has a nullable EXTRA column while mysql has a non-nullable